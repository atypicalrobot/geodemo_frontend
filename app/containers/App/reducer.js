/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_STORIES_SUCCESS,
  LOAD_STORIES,
  LOAD_STORIES_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentGenre: false,
  storyData: {
    stories: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_STORIES:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['storyData', 'stories'], false);
    case LOAD_STORIES_SUCCESS:
      return state
        .setIn(['storyData', 'stories'], action.stories)
        .set('loading', false)
        .set('currentGenre', action.genre);
    case LOAD_STORIES_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
