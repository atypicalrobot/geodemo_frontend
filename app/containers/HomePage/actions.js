/*
 * Home Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  CHANGE_GENRE,
  CHANGE_STORY,
  CHANGE_POI,
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {name} name The new text of the input field
 *
 * @return {object}    An action object with a type of CHANGE_GENRE
 */
export function changeGenre(genre) {
  return {
    type: CHANGE_GENRE,
    genre,
  };
}

export function changeStory(story) {
  return {
    type: CHANGE_STORY,
    story,
  };
}

export function changePoi(poi) {
  return {
    type: CHANGE_POI,
    poi,
  };
}
