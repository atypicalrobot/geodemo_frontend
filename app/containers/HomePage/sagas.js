/**
 * Gets the stories of the user from Github
 */

import { take, call, put, select, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_STORIES } from 'containers/App/constants';
import { storiesLoaded, storyLoadingError } from 'containers/App/actions';

import request from 'utils/request';
import { makeSelectGenre } from 'containers/HomePage/selectors';

/**
 * Github stories request/response handler
 */
export function* getStories() {
  // Select genre from store
  const genre = yield select(makeSelectGenre());
  const requestURL = `https://api.github.com/users/${genre}/stories?type=all&sort=updated`;

  try {
    // Call our request helper (see 'utils/request')
    const stories = yield call(request, requestURL);
    yield put(storiesLoaded(stories, genre));
  } catch (err) {
    yield put(storyLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* githubData() {
  // Watches for LOAD_STORIES actions and calls getStories when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_STORIES, getStories);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  githubData,
];
