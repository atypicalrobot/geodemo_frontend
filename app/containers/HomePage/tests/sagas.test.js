/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_STORIES } from 'containers/App/constants';
import { storiesLoaded, storyLoadingError } from 'containers/App/actions';

import { getStories, githubData } from '../sagas';

const genre = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('getStories Saga', () => {
  let getStoriesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getStoriesGenerator = getStories();

    const selectDescriptor = getStoriesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getStoriesGenerator.next(genre).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the storiesLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First story',
    }, {
      name: 'Second story',
    }];
    const putDescriptor = getStoriesGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(storiesLoaded(response, genre)));
  });

  it('should call the storyLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getStoriesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(storyLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_STORIES action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_STORIES, getStories));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = githubDataSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = githubDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
