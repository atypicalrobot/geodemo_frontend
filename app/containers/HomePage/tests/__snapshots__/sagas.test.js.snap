exports[`getStories Saga should call the storyLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getStories Saga should call the storyLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/stories?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getStories Saga should dispatch the storiesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getStories Saga should dispatch the storiesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/stories?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
